test/integration/integration.test.ts

```diff
@@ -0,0 +1,678 @@
+import { describe, it } from 'vitest';
+import { setTimeout } from 'node:timers/promises';
+import { createClients } from './utilities/setup.js';
+import { LocationsEventMap } from '../../src/Locations.js';
+import { SpaceEventMap } from '../../src/Space.js';
+import { MembersEventMap } from '../../src/Members.js';
+import { CursorsEventMap } from '../../src/Cursors.js';
+import { nanoid } from 'nanoid';
+import { LocksEventMap } from '../../src/Locks.js';
+
+describe.concurrent(
+  'integration tests',
+  () => {
+    it('space members', async (context) => {
+      // Given Spaces clients `performerSpaces` and `observerSpaces`, both configured to use the same API key, and each configured to use a different randomly-generated client ID...
+      const [{ spaces: performerSpaces, clientId: performerClientId }, { spaces: observerSpaces }] =
+        await createClients({ count: 2 });
+
+      // ...and Space instances `performerSpace` and `observerSpace`, fetched from `performerSpaces` and `observerSpaces` respectively using `#get` with a randomly-generated space name, and passing an `offlineTimeout` option of 5 seconds when creating `observerSpace`...
+      //
+      // (Motivation for choosing 5 seconds for `offlineTimeout`: in scenario_1_4 we wait to observe the members `remove` event, which requires us to wait a duration of `offlineTimeout`, and hence means that execution of this test case will take at least `offlineTimeout`. We choose 5 seconds as an arbitrary value which is small enough such that we’re happy for the test to take this long to execute, but not so short as to run the risk of missing this `remove` event whilst waiting for scenario_1_3’s `leave()` to return.)
+      const spaceName = nanoid();
+      const performerSpace = await performerSpaces.get(spaceName);
+      const observerSpace = await observerSpaces.get(spaceName, { offlineTimeout: 5000 });
+
+      // (scenario_1_1: entering a space)
```

Comment:
I'm afraid I'm not sure exactly what it is that you're suggesting here. What it _sounds_ like is that, given a single test that’s written linearly like this:

```
scenario_1:
  doSomething_1
  checkSomething_1

scenario_2:
  doSomething_2
  checkSomething_2

scenario_3:
  doSomething_3
  checkSomething_3
```

then you're suggesting that we instead write it as nested test cases something like this?

```
describe('scenario_1', () => {
  beforeEach(() => {
    // do the actions of doSomething_1 and somehow expose a way for the `it` block to check its results
  });

  it('some description of scenario_1', () => {
    // do checkSomething_1
  });

  describe('scenario_2', () => {
    beforeEach(() => {
      // do the actions of doSomething_2 and somehow expose a way for the below `it` block to check its results
    });

    it('some description of scenario_2', () => {
      // do checkSomething_2
    });

    describe('scenario_3', () => {
      beforeEach(() => {
        // do the actions of doSomething_3 and somehow expose a way for the below `it` block to check its results
      });

      it('some description of scenario_3', () => {
        // do checkSomething_3
      });
    });
  });
});
````

Have I got that right? If so, a couple of things that I'd wonder about:

- isn't this code going to get very indented, very quickly?
- I'm not sure how it achieves "instead of having to parse the whole test to debug it" — given that each scenario relies on the side effects given in the preceding scenarios, don't you still have to understand all of the preceding sections in order to confidently edit a given section?

If I've misunderstood what you're suggesting, would you mind perhaps showing a concrete example of what you had in mind?

Reply:
> isn't this code going to get very indented, very quickly?

this is how the unit tests are structured, I'm not sure why that's an issue? (I might be biased - for me this is pretty standard in JavaScript and Ruby ecosystems).

> I'm not sure how it achieves "instead of having to parse the whole test to debug it" — given that each scenario relies on the side effects given in the preceding scenarios, don't you still have to understand all of the preceding sections in order to confidently edit a given section?

with your current structure, if a test fails it's something like:
```
cursors test failed - failed
// expectation which failed
```

So we need to go over all the lines of code up to the point that a given expectation has failed. Compare that to:

```
cursors test
  assertion description and setup 1 - passed
  assertion description and setup 2 - passed
  assertion description and setup 3 - failed
```

Without looking at code or expectations, we can see what assertion, assumption fails, and also which ones are still good. With a linear setup, it's likely no assertion passes from the first failure, but that's fine. This is how Cucumber specs work in Ruby.

Reply:
Just adding a comment to record that we had a chat about this and it turns out I was a bit thrown by your usage of the term "nested", where you actually just meant to keep the structure of the tests as-is but to nest each test case in a `describe`, put the setup part into a `beforeAll` and each scenario into its own `it`. I've made those changes and will push them shortly.